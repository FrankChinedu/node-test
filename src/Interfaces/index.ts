/* eslint-disable @typescript-eslint/no-explicit-any */

import { type } from "os"

export interface IResponse {
  token?: string;
  status: number;
  success: boolean;
  data?: any;
  error?: any,
  message?: string
}

export type UserDataType = {
  name: string,
  email: string,
  password: string
}

export interface UserType {
  id: number,
  email: string,
  name: string
}
