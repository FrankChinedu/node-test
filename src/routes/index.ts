import { Express } from "express";
import user from "./user";
import post from "./post";

export default (app: Express): void => {
  app.use("/api/v1", user);
  app.use("/api/v1/post", post);
};
