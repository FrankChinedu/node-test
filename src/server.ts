import express from "express";
import { json } from "body-parser";
import cors from "cors";
import consola from "consola";
import swaggerUi from 'swagger-ui-express';
import YAML from 'yamljs';

import Route from "./routes/index";

import { APP_PORT } from "./config/env";

const app = express();

//API DOCS
const swaggerDocument = YAML.load(`${__dirname}/swagger.yaml`);
app.use('/api-docs', swaggerUi.serve, swaggerUi.setup(swaggerDocument));

app.use(json());
app.use(cors());

app.get("/", (req, res) => {
  res.json({
    success: true,
    message: "Welcome ðŸ˜ƒ",
  });
});

Route(app);

app.use((req, res) => res.status(404).json({ error: "Route not found" }));

app.listen(APP_PORT, () => {
  consola.success(`server is listening on port ${APP_PORT}`);
});

export { app };
