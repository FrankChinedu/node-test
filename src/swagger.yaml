swagger: '2.0'
info:
  version: '1.0'
  title: post-like
  contact: {}
host: example.com
basePath: /
securityDefinitions: {}
schemes:
- http
consumes:
- application/json
produces:
- application/json
paths:
  /api/v1/register:
    post:
      summary: register user
      tags:
      - user
      operationId: registeruser
      deprecated: false
      produces:
      - application/json
      parameters:
      - name: Authorization
        in: header
        required: false
        default: Bearer {token}
        type: string
      - name: Body
        in: body
        required: true
        description: ''
        schema:
          $ref: '#/definitions/registeruserrequest'
      responses:
        200:
          description: ''
          headers: {}
  /api/v1/login:
    post:
      summary: login user
      tags:
      - user
      operationId: loginuser
      deprecated: false
      produces:
      - application/json
      parameters:
      - name: Authorization
        in: header
        required: false
        default: Bearer {token}
        type: string
      - name: Body
        in: body
        required: true
        description: ''
        schema:
          $ref: '#/definitions/loginuserrequest'
      responses:
        200:
          description: ''
          headers: {}
  /api/v1/user:
    patch:
      summary: update user
      tags:
      - user
      operationId: updateuser
      deprecated: false
      produces:
      - application/json
      parameters:
      - name: Authorization
        in: header
        required: false
        default: Bearer {token}
        type: string
      - name: Body
        in: body
        required: true
        description: ''
        schema:
          $ref: '#/definitions/updateuserrequest'
      responses:
        200:
          description: ''
          headers: {}
    get:
      summary: get self
      tags:
      - user
      operationId: getself
      deprecated: false
      produces:
      - application/json
      parameters:
      - name: Authorization
        in: header
        required: false
        default: Bearer {token}
        type: string
      responses:
        200:
          description: ''
          headers: {}
    delete:
      summary: delete self
      tags:
      - user
      operationId: deleteSelf
      deprecated: false
      produces:
      - application/json
      parameters:
      - name: Authorization
        in: header
        required: false
        default: Bearer {token}
        type: string
      responses:
        200:
          description: ''
          headers: {}
  /api/v1/post:
    get:
      summary: get posts
      tags:
      - post
      operationId: getposts
      deprecated: false
      produces:
      - application/json
      parameters:
      - name: Authorization
        in: header
        required: false
        default: Bearer {token}
        type: string
      - name: page
        in: query
        required: true
        type: integer
        format: int32
        description: ''
      - name: limit
        in: query
        required: true
        type: integer
        format: int32
        description: ''
      responses:
        200:
          description: ''
          headers: {}
    post:
      summary: create post
      tags:
      - post
      operationId: createpost
      deprecated: false
      produces:
      - application/json
      parameters:
      - name: Authorization
        in: header
        required: false
        default: Bearer {token}
        type: string
      - name: Body
        in: body
        required: true
        description: ''
        schema:
          $ref: '#/definitions/createpostrequest'
      responses:
        200:
          description: ''
          headers: {}
  /api/v1/post/{id}:
    put:
      summary: update post
      tags:
      - post
      operationId: updatepost
      deprecated: false
      produces:
      - application/json
      parameters:
      - name: Authorization
        in: header
        required: false
        default: Bearer {token}
        type: string
      - name: id
        in: path
        required: true
        type: integer
        format: int32
        description: ''
      - name: Body
        in: body
        required: true
        description: ''
        schema:
          $ref: '#/definitions/updatepostrequest'
      responses:
        200:
          description: ''
          headers: {}
    get:
      summary: get one post
      tags:
      - post
      operationId: getonepost
      deprecated: false
      produces:
      - application/json
      parameters:
      - name: Authorization
        in: header
        required: false
        default: Bearer {token}
        type: string
      - name: id
        in: path
        required: true
        type: integer
        format: int32
        description: ''
      responses:
        200:
          description: ''
          headers: {}
    delete:
      summary: delete post
      tags:
      - post
      operationId: deletepost
      deprecated: false
      produces:
      - application/json
      parameters:
      - name: Authorization
        in: header
        required: false
        default: Bearer {token}
        type: string
      - name: id
        in: path
        required: true
        type: integer
        format: int32
        description: ''
      responses:
        200:
          description: ''
          headers: {}
  /:
    get:
      summary: home
      tags:
      - Misc
      operationId: home
      deprecated: false
      produces:
      - application/json
      parameters:
      - name: Authorization
        in: header
        required: false
        default: Bearer {token}
        type: string
      responses:
        200:
          description: ''
          headers: {}
definitions:
  registeruserrequest:
    title: registeruserrequest
    example:
      name: franka ang
      email: franka@me.com
      password: 123456
    type: object
    properties:
      name:
        type: string
      email:
        type: string
      password:
        type: string
    required:
    - name
    - email
    - password
  loginuserrequest:
    title: loginuserrequest
    example:
      email: franka@me.com
      password: 123456
    type: object
    properties:
      email:
        type: string
      password:
        type: string
    required:
    - email
    - password
  updateuserrequest:
    title: updateuserrequest
    example:
      name: lo frank
    type: object
    properties:
      name:
        type: string
    required:
    - name
  createpostrequest:
    title: createpostrequest
    example:
      name: 'again '
      body: again logn sop
    type: object
    properties:
      name:
        type: string
      body:
        type: string
    required:
    - name
    - body
  updatepostrequest:
    title: updatepostrequest
    example:
      name: new name
      body: ' an new logn body'
    type: object
    properties:
      name:
        type: string
      body:
        type: string
    required:
    - name
    - body
tags:
- name: user
- name: post
- name: Misc
  description: ''
