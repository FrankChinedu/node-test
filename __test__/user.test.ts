import supertest from 'supertest';
import { app } from '../src/server';

const server = () => supertest(app);

const user =  {
  name: "franka ang",
  email: "franka@me.com",
  password: "123456"
}

describe('User', () => {
  it('should create a new user', async () => {
    const { status, body } = await server().post('/api/v1/register').send(user);
    expect(status).toBe(200);
    expect(body.status).toBe(200);
    const bodyKeys = ['success', 'status', 'data', 'token'];
    expect(Object.keys(body)).toEqual(bodyKeys);
    const dataKeys = ['id', 'name', 'email'];
    expect(Object.keys(body.data)).toEqual(dataKeys);
  });

  it('should throw error if user already exist', async () => {
    await server().post('/api/v1/register').send(user);
    const { status, body } = await server().post('/api/v1/register').send(user);
    expect(status).toBe(500);
    expect(body.status).toBe(500);
    const bodyKeys = ['success', 'status', 'error'];
    expect(Object.keys(body)).toEqual(bodyKeys);
  });

  it('should return validtion error if some body params are missing', async () => {
    
    const { status, body } = await server().post('/api/v1/register').send({
      name: user.name,
      email: user.email
    });
    expect(status).toBe(400);
    expect(body).toMatchSnapshot();
  });

  it('user can login', async () => {

    await server().post('/api/v1/register').send(user);
    const { status, body } = await server().post('/api/v1/login').send({
      email: user.email,
      password: user.password
    });
    expect(status).toBe(200);
    const bodyKeys = ['success', 'status', 'data', 'token'];
    expect(Object.keys(body)).toEqual(bodyKeys);
    const dataKeys = ['id', 'name', 'email'];
    expect(Object.keys(body.data)).toEqual(dataKeys);
  });

});
