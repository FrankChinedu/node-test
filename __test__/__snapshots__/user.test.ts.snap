// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`User should return validtion error if some body params are missing 1`] = `
Object {
  "error": Object {
    "_original": Object {
      "email": "franka@me.com",
      "name": "franka ang",
    },
    "details": Array [
      Object {
        "context": Object {
          "key": "password",
          "label": "password",
        },
        "message": "\\"password\\" is required",
        "path": Array [
          "password",
        ],
        "type": "any.required",
      },
    ],
  },
  "success": false,
}
`;
